segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -552,36 +559,173 @@ private CompletableFuture<Void> writeSnapshotInfo(long snapshotId) {
                                 pendingGarbageChunks.clear();
                             }, executor)
                             .exceptionally(e -> {
-                                log.error("Unable to persist snapshot info.{}", currentSnapshotIndex, e);
+                                log.error("SystemJournal[{}] Unable to persist snapshot info.{}", containerId, currentSnapshotIndex, e);
                                 return null;
                             });
                 }, executor);
     }
 
+    /**
+     * Writes SnapshotInfo to a well known file.
+     */
+    private CompletableFuture<Void> writeSnapshotInfoToFile(SnapshotInfo info) {
+        val snapshotInfoFileName = NameUtils.getSystemJournalSnapshotInfoFileName(containerId);
+        try {
+            val isDone = new AtomicBoolean(false);
+            val attempts = new AtomicInteger();
+            val bytes = SNAPSHOT_INFO_SERIALIZER.serialize(info);
+
+            return Futures.loop(
+                    () -> !isDone.get(),
+                    () -> {
+                        return chunkStorage.exists(snapshotInfoFileName)
+                                .thenComposeAsync( exists -> {
+                                    if (exists) {
+                                        // Read back and make sure higher epoch has not already written it
+                                        return readSnapshotInfoFromFile()
+                                                .thenComposeAsync( existingSnapshot -> {
+                                                    if (existingSnapshot.getEpoch() > epoch) {
+                                                        return CompletableFuture.failedFuture(new StorageNotPrimaryException(String.format("SystemJournal[{}] Unexpected snapshot. Expected = {} actual = {}", info, existingSnapshot)));
+                                                    } else {
+                                                        return chunkStorage.delete(ChunkHandle.writeHandle(snapshotInfoFileName));
+                                                    }
+                                                }, executor);
+                                    } else {
+                                        return CompletableFuture.completedFuture(null);
+                                    }
+                                }, executor)
+                                .thenComposeAsync(v -> chunkStorage.createWithContent(snapshotInfoFileName,
+                                                bytes.getLength(),
+                                                new ByteArrayInputStream(bytes.array(), bytes.arrayOffset(), bytes.getLength())),
+                                        executor)
+                                .thenAcceptAsync(v -> {
+                                    log.debug("SystemJournal[{}] Snapshot info saved to LTS. File {}. info = {}", containerId, snapshotInfoFileName, info);
```

Comment:
nit: This can be combined with the next step to avoid the extra .thenAcceptAsync

Reply:
updated